### PR TITLE
Add extended scheduler to operator

### LINKED ISSUE #138: Extend scheduler for better support local persistent volume
We can't schedule the 2nd pod(pd/tikv) on the same node as the 1st pod, nor schedule the 3rd pod on the same node as the first two pods.

So we should add an [extended scheduler](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/scheduler_extender.md) to the operator to achieve it.

### PR DESCRIPTION
**Pull request by weekface**
This PR adds an extended scheduler to TiDB Operator, and fixes #138:

- Basic extended scheduler framework
- HighAvailability predicate
- Unit tests
- E2E tests

@tennix @onlymellb @xiaojingchen @gregwebs PTAL

### CODE REVIEW

File: pkg/scheduler/predicates/ha.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		}
+
+		if ordinal1 < ordinal && nodeName == "" {
+			return nil, fmt.Errorf("waiting for pod: %s/%s to be scheduled", ns, podName1)
+		}
+		if nodeName == "" {
+			continue
+		}
+
+		nodeMap[nodeName] = append(nodeMap[nodeName], podName1)
```
Comment (xiaojingchen) [inline pkg/scheduler/predicates/ha.go]: if the nodeName is not contain in nodeMap, there will happen panic ,we should add logic here
```
if nodeMap[nodeName] ==nil{
      continue
}
```

File: pkg/scheduler/predicates/ha.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	var min int
+	var minInitialized bool
```
Comment (xiaojingchen) [inline pkg/scheduler/predicates/ha.go]: the `minInitialized` is not used,what is the `minInitialized` mean

File: pkg/scheduler/predicates/ha.go (hunk 1 above)
Comment (weekface) [reply pkg/scheduler/predicates/ha.go]: Good catch!
Done!

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+// else return the original nodes.
+func (s *scheduler) Filter(args *schedulerapiv1.ExtenderArgs) (*schedulerapiv1.ExtenderFilterResult, error) {
+	pod := &args.Pod
+	ns := pod.GetNamespace()
+	podName := pod.GetName()
+	kubeNodes := args.Nodes.Items
+
+	var clusterName string
+	var exist bool
+	if clusterName, exist = pod.Labels[label.InstanceLabelKey]; !exist {
```
Comment (tennix) [inline pkg/scheduler/scheduler.go]: I think returning the original nodes and logging a message is more reasonable in this case.

File: pkg/scheduler/scheduler.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+		return &schedulerapiv1.ExtenderFilterResult{
+			Nodes:     args.Nodes,
+			NodeNames: args.NodeNames,
+		}, nil
+	}
+
+	glog.Infof("scheduling pod: %s/%s", ns, podName)
+	var err error
+	for _, predicate := range s.predicates {
+		glog.Infof("entering predicate: %s, nodes: %v", predicate.Name(), getNodeNames(kubeNodes))
```
Comment (tennix) [inline pkg/scheduler/scheduler.go]: This log should have a higher log level since it's much more verbose.

File: pkg/scheduler/scheduler.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	return &scheduler{
+		kubeCli: kubeCli,
+		predicates: []predicates.Predicate{
+			predicates.NewHA(kubeCli, pdReplicas, tikvReplicas),
+		},
+	}
+}
+
+// Filter selects a set of nodes from *schedulerapiv1.ExtenderArgs.Nodes when this is a pd or tikv pod
+// else return the original nodes.
```
Comment (tennix) [inline pkg/scheduler/scheduler.go]: otherwise, returns the original nodes.

File: pkg/scheduler/scheduler.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+		}
+		glog.Infof("leaving predicate: %s, nodes: %v", predicate.Name(), getNodeNames(kubeNodes))
+	}
+
+	return &schedulerapiv1.ExtenderFilterResult{
+		Nodes: &apiv1.NodeList{Items: kubeNodes},
+	}, nil
+}
+
+// We didn't pass `prioritizeVerb` to kubernetes scheduler extender's config file, this method will not be called.
```
Comment (tennix) [inline pkg/scheduler/scheduler.go]: s/didn't/don't/

File: pkg/scheduler/predicates/ha.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+func (h *ha) Filter(clusterName string, pod *apiv1.Pod, nodes []apiv1.Node) ([]apiv1.Node, error) {
+	ns := pod.GetNamespace()
+	podName := pod.GetName()
+
+	var component string
+	var exist bool
+	if component, exist = pod.Labels[label.ComponentLabelKey]; !exist {
+		return nil, fmt.Errorf("can't find component in pod labels: %s/%s", ns, podName)
```
Comment (tennix) [inline pkg/scheduler/predicates/ha.go]: return the nodes as it is.

File: pkg/scheduler/predicates/ha.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	var replicas int32
+	switch component {
+	case label.PDLabelVal:
+		replicas = h.pdReplicas
+	case label.TiKVLabelVal:
+		replicas = h.tikvReplicas
+	}
+
+	if ordinal < replicas && min != 0 {
+		return nil, fmt.Errorf("the first %d pods can't scheduled to the same node", replicas)
```
Comment (tennix) [inline pkg/scheduler/predicates/ha.go]: can't be scheduled

File: charts/tidb-operator/templates/scheduler-policy-configmap.yaml [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  name: tidb-scheduler-policy
+  labels:
+    app.kubernetes.io/name: {{ template "chart.name" . }}
+    app.kubernetes.io/managed-by: {{ .Release.Service }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    app.kubernetes.io/component: scheduler
+    helm.sh/chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+"  "_" }}
+data:
+  policy.cfg: |-
+    {
```
[the remark this answers is not in the document]
Comment (weekface) [reply charts/tidb-operator/templates/scheduler-policy-configmap.yaml]: OK, waiting for: https://github.com/pingcap/tidb-operator/pull/149 merged.

File: images/tidb-operator-e2e/tidb-operator-values.yaml [hunk 10]
```diff
@@ -29,3 +35,25 @@ controllerManager:
...
+      cpu: 250m
+      memory: 150Mi
+    requests:
+      cpu: 80m
+      memory: 50Mi
+  # pd replicas
+  pdReplicas: 3
+  # tikv replicas
+  tikvReplicas: 3
+  hyperkubeImage: quay.io/coreos/hyperkube:v1.10.4_coreos.0
```
Comment (tennix) [inline images/tidb-operator-e2e/tidb-operator-values.yaml]: Use `kubeSchedulerImage` instead.

File: pkg/scheduler/predicates/ha.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+	}
+	h.podListFn = h.realPodListFn
+	return h
+}
+
+func (h *ha) Name() string {
+	return "HighAvailability"
+}
+
+func (h *ha) Filter(clusterName string, pod *apiv1.Pod, nodes []apiv1.Node) ([]apiv1.Node, error) {
```
Comment (tennix) [inline pkg/scheduler/predicates/ha.go]: Add a short description of the algrithom how we schedule pods.

1. First, we sort all the nodes we get from kube-scheduler by how many same kind of pod it contains, find the nodes that have least pods.
2. When scheduling the first `replicas` pods, we must ensure no previous pods on the nodes.
3. For later pods, we choose the nodes that have least pods.

**Pull request merged**